### PR TITLE
schema tests defined by macros

### PR DESCRIPTION
**Pull request by drewbanin**
- use "globally" scoped macros instead of hardcoded sql in python files
- supports user-defined schema tests (via macros) https://github.com/analyst-collective/dbt/issues/200

To create a custom schema test, create macros prefixed with `test_`. The `test_` macros *must* take `model` as the first argument. If the test takes a single argument (as with not null & unique tests), then the second argument should be called `arg`. Otherwise, the remaining arguments can be named as you like. Test configurations stated in `schema.yml` will be mapped to the macro arguments as long as the names match. For example:

```sql
-- macros/tests.sql
{% macro test_something(model, arg) %}

select *
from {{ ref(model) }}

where id = '{{ arg }}'

{% endmacro %}


{% macro test_something_complicated(model, first_argument, another_argument) %}

select *
from {{ ref(model) }}

where id = '{{ first_argument }}'
   or id = '{{ another_argument }}'

{% endmacro %}
```

```yml
some_model:
  constraints:
    my_project.test_something:
      - abc
      - def

    my_project.test_something_complicated:
      - {first_argument: abc, another_argument: def}
```

### CODE REVIEW

File: dbt/parser.py [hunk 1]
```diff
@@ -341,57 +304,52 @@ def parse_schema_tests(tests, root_project, projects):
...
 
-    elif test_type == 'relationships':
-        if not isinstance(test_config, dict):
-            return None
+        if type(arg_val) == dict:
+            parts = arg_val.values()
+        elif type(arg_val) in (list, tuple):
+            parts = arg_val
+        else:
+            parts = [arg_val]
```
Comment (cmcarthur) [inline dbt/parser.py]: can this be anything but a dict? looks like below we make sure it's a dict before calling this. `parts = kwargs.values()` should cover all the cases, right?

File: dbt/parser.py [hunk 2]
```diff
@@ -341,57 +304,52 @@ def parse_schema_tests(tests, root_project, projects):
...
 
-        child_field = test_config.get('from')
-        parent_field = test_config.get('field')
-        parent_model = test_config.get('to')
+        flat_args.extend([str(part) for part in parts])
```
Comment (cmcarthur) [inline dbt/parser.py]: use dbt.compat.basestring

File: dbt/utils.py [hunk 3]
```diff
@@ -152,21 +162,31 @@ def find_model_by_fqn(models, fqn):
...
-                    project.profiles_dir,
-                    profile_to_load=project.profile_to_load,
-                    args=project.args)
-            except dbt.project.DbtProjectError as e:
-                logger.info(
-                    "Error reading dependency project at {}".format(full_obj)
-                )
-                logger.info(str(e))
+def dependency_projects(project, include_global=True):
+    if include_global:
```
Comment (cmcarthur) [inline dbt/utils.py]: is there a time when we would not want to include globals?

File: dbt/parser.py (hunk 1 above)
Comment (drewbanin) [inline dbt/parser.py]: This func takes arbitrary schema test configs and spits out a human readable, unique name.

```yml
my_model:
  constraints:
    test_something:
      - { some_value: True, other_thing: ['abc', 'def'] }
```

So this func would operate on `{ some_value: True, other_thing: ['abc', 'def'] }` and spit out:

```
test_something_my_model_abc_def__True
```

I actually think that in practice this is kind of annoying/confusing. This "nice" name becomes the compiled filename and shows up in the `dbt test` output. Ideally, we'd keep the test config args around and show something like:

```
ERROR running test "test_something" for model "my_model" with args:
  some_value: True
  other_thing: ['abc', 'def']
```

but that's not really how things work currently. Something to consider for the future though

Comment (drewbanin) [inline dbt/parser.py]: oh gosh, and to answer your question, yes: The  `arg_val` var will be the value of each item in the supplied dict (`args`).

File: dbt/parser.py (hunk 2 above)
Comment (drewbanin) [inline dbt/parser.py]: :+1:

File: dbt/utils.py (hunk 3 above)
Comment (drewbanin) [inline dbt/utils.py]: nope, i'll remove this

File: dbt/parser.py [hunk 4]
```diff
@@ -341,57 +306,61 @@ def parse_schema_tests(tests, root_project, projects):
...
-        parent_model = test_config.get('to')
 
-        raw_sql = QUERY_VALIDATE_REFERENTIAL_INTEGRITY.format(
-            child_field=child_field,
-            child_ref="{{ref('"+model_name+"')}}",
-            parent_field=parent_field,
-            parent_ref=("{{ref('"+parent_model+"')}}"))
+def as_kwarg(key, value):
+    test_value = to_string(value)
+    is_function = re.match(r'^\s*(ref|var)\(.+\)$', test_value) is not None
```
Comment (drewbanin) [inline dbt/parser.py]: @cmcarthur how do you feel about this?

Comment (cmcarthur) [inline dbt/parser.py]: i feel ok about it. not in love with regex parsing but what can you do.

is the intention to support ref and var in custom schema tests? it doesn't look like those would be passed in here unless i'm missing something

File: test/integration/008_schema_tests_test/models/schema.yml [hunk 5]
```diff
@@ -29,7 +29,7 @@ table_summary:
             - { field: favorite_color, values: ['blue', 'green'] }
 
         relationships:
-            - { from: favorite_color, to: table_copy, field: favorite_color }
+            - { from: favorite_color, to: ref('table_copy'), field: favorite_color }
```
Comment (cmcarthur) [inline test/integration/008_schema_tests_test/models/schema.yml]: ahh i understand now. is this a breaking change in how schema tests are defined?

**Comment by drewbanin** [comment]
Yeah, sorry for the lack of context. It is a breaking change, but we don't really have much of a choice! I think it's more canonical like this anyway -- kind of weird that the ref was implicit before. I ran it by @jthandy and he agreed

**Pull request closed without merging**